net/host.go

```diff
@@ -308,86 +334,113 @@ func (h *host) addrInfo() peer.AddrInfo {
 	}
 }
 
-func (h *host) writeToStream(s libp2pnetwork.Stream, msg Message) error {
+func writeStreamMessage(s io.Writer, msg Message, peerID peer.ID) error {
 	encMsg, err := msg.Encode()
 	if err != nil {
 		return err
 	}
 
-	msgLen := uint64(len(encMsg))
-	lenBytes := uint64ToLEB128(msgLen)
-	encMsg = append(lenBytes, encMsg...)
+	err = binary.Write(s, binary.LittleEndian, uint32(len(encMsg)))
+	if err != nil {
+		return err
+	}
 
 	_, err = s.Write(encMsg)
 	if err != nil {
 		return err
 	}
 
-	log.Debug(
-		"Sent message to peer=", s.Conn().RemotePeer(), " type=", msg.Type(),
-	)
+	log.Debugf("Sent message to peer=%s type=%s", peerID, msg.Type())
 
 	return nil
 }
 
-// readStream reads from the stream into the given buffer, returning the number of bytes read
-func readStream(stream libp2pnetwork.Stream, buf []byte) (int, error) {
-	if stream == nil {
-		return 0, errNilStream
+// readStreamMessage reads the 4-byte LE size header and message body returning the
+// message body bytes. io.EOF is returned if the stream is closed before any bytes
+// are received. If a partial message is received before the stream closes,
+// io.ErrUnexpectedEOF is returned.
+func readStreamMessage(s io.Reader) (Message, error) {
+	if s == nil {
+		return nil, errNilStream
 	}
 
-	var (
-		tot int
-	)
-
-	length, bytesRead, err := readLEB128ToUint64(stream, buf[:1])
+	lenBuf := make([]byte, 4) // uint32 size
+	n, err := io.ReadFull(s, lenBuf)
 	if err != nil {
-		return bytesRead, fmt.Errorf("failed to read length: %w", err)
-	}
-
-	if length == 0 {
-		return 0, nil
+		if isEOF(err) {
+			if n > 0 {
+				err = io.ErrUnexpectedEOF
+			} else {
+				err = io.EOF
+			}
+		}
+		return nil, err
 	}
+	msgLen := binary.LittleEndian.Uint32(lenBuf)
 
-	if length > uint64(len(buf)) {
-		log.Warnf("received message with size greater than allocated message buffer: msg size=%d, buffer size=%d",
-			length, len(buf))
-		return 0, fmt.Errorf("message size greater than allocated message buffer: got %d", length)
+	if msgLen > maxMessageSize {
+		log.Warnf("Received message longer than max allowed size: msg size=%d, max=%d",
+			msgLen, maxMessageSize)
+		return nil, fmt.Errorf("message size %d too large", msgLen)
 	}
 
-	tot = 0
-	for i := 0; i < maxReads; i++ {
-		n, err := stream.Read(buf[tot:])
-		if err != nil {
-			return n + tot, err
-		}
-
-		tot += n
-		if tot == int(length) {
-			break
+	msgBuf := make([]byte, msgLen)
```

Comment:
imo it's a bad idea to allocate buffers based on a peer opening a stream as it's a DOS vector. we should fix this in a follow up

Reply:
I agree that it is a potential DDOS issue.  We don't hold on to the buffers long, and if the message is bad, we close the connection. I didn't think these changes were a degradation from what the code was already doing, as we were already allocating the buffer on every new stream here:
https://github.com/AthanorLabs/atomic-swap/blob/master/net/initiate.go#L62

Reply:
yeah, I think this is fine for this PR but should be fixed still. for example a peer could connect to us an open as many `query` streams as possible (since we can have multiple streams open with a peer with the same protocol id), which would result in us going OOM.

Reply:
We should definitely put some time into testing/attempting/preventing DDOS attacks. I'm trying to test your DDOS idea right now, but I can't open more than 256 streams on the same connection. It looks like the max number of streams is configurable per direction:
https://github.com/libp2p/go-libp2p/blob/master/p2p/host/resource-manager/limit.go#L25-L26

Reply:
good to know, maybe just limiting the number of open streams a peer can have would be okay, since we have 50 max peers. if we limit open streams per peer to say, 10, the worst that can happen is they force us to do many large allocations in sequentially (which is still bad, but probably won't cause OOM?)

Reply:
I think we'd be fine. The max buffer size right now is "1 << 17" (131072).
```
(50 peers * 10 connections * 131072 bytes) / (1024 * 1024) = 62.5 MB
```
Because the streams are multiplexed over the same connection, and we would only be reading one message at a time, we are probably no longer referencing the previous buffer when reading the next one (i.e. GC can recycle it).  libp2p is limiting how much unread data can be cached on the receiver's stream as well.